feat: anon piece selection and retrieval (#487)
FilOzone/dealbot: docs/checks/events-and-metrics.md

This document describes the expected flow and metrics. Items marked **TBD** are not yet implemented but will get reviewed and cleaned up as part of https://github.com/FilOzone/dealbot/issues/280.

> ## Anonymous Retrieval Event Model

Reviewer: I think we can move this section after "Data Storage Event Model"

The [Anonymous Retrieval check](./anon-retrievals.md) is a single-shot flow per piece: select → fetch piece → (optional) parse CAR + IPNI + block fetch → write one ClickHouse row.

> It is not modeled as a sequence of named lifecycle events. Instead it emits:

Reviewer: It wasn't clear to me why this should be any different from a presentation regard given:
- It also has mulitiple "actors" (dealbot, SP, subgraph, filecoinpin.contact. (I guess we could even add Clickhouse metric emitting not essential)
- We do have events. `anonPieceRetrievalFirstByteMs` references "Piece fetch start" and "First byte received from /piece/{pieceCid}". Ideally we would be hyperlinking to something like `anonPieceRetrievalStart`, `anonPieceRetrievalFirstByteReceived` as examples

Reviewer: I think this is still relevant. Right now the document reads weird with two "Data Storage Event Model" sections.

I'm suggesting we follow something similar to what was recently done for https://github.com/filozone/dealbot/blob/main/docs/checks/events-and-metrics.md#pull-check-event-model
1. Add "Anon Retrieval Check Event Timeline"
2. Add "Anon Retrieval Check Event List"
3. Update "Metrics" section with emetrics (like you ahve done) but then tie back with anchor links to the specific events. Example: https://github.com/filozone/dealbot/blob/main/docs/checks/events-and-metrics.md#pullRequestAcknowledgementLatencyMs

> - **Outcome metrics** when each step completes — see the [time](#time-related-metrics) and [status](#status-count-related-metrics) metric tables for `anonPieceRetrievalFirstByteMs`, `anonRetrievalCheckMs`, `anonRetrievalStatus`, `anonCarParseStatus`, `anonIpniStatus`, `anonBlockFetchStatus`, and friends.

Reviewer: If we do adopt this format rather than the format that is here for data storage check (and for the new pull check https://github.com/FilOzone/dealbot/blob/629be23b8e7996d1788a9d9fa94f1cd753398539/docs/checks/events-and-metrics.md#pull-check-event-model), then lets have all of these be relative hyperlinks for easy human navigation.

| <a id="ipfsRetrievalIntegrityChecked"></a>`ipfsRetrievalIntegrityChecked` | Retrieved content matches expected CID (per-block sha256 hash verification via `createBlock`). Inline check at end of DAG traversal; no discrete event emission. | Data Storage, Retrieval | [`ipfs-block.strategy.ts`](../../apps/backend/src/retrieval-addons/strategies/ipfs-block.strategy.ts) |

## Anon Retrieval Check Event Model

Below are the events for an [Anonymous Retrieval check](./anon-retrievals.md). Unlike the Data Storage flow, anonymous retrieval is a single-shot per-piece flow: select a publicly-discoverable piece from the FWSS subgraph → fetch the piece from the SP → optionally parse the CAR, verify IPNI, and sample block fetches → write one row to ClickHouse. The check emits one row to `anon_retrieval_checks` even on abort or unexpected error, so partial timing/byte evidence is never lost.

### Anon Retrieval Check Event Timeline

```mermaid

> sequenceDiagram

Reviewer: Nice - looks good.

| <a id="anonPieceFetchLastByteReceived"></a>`anonPieceFetchLastByteReceived` | Last byte received from `/piece/{pieceCid}`.                                                                                                                                                                         | [`piece-retrieval.service.ts`](../../apps/backend/src/retrieval-anon/piece-retrieval.service.ts) (drives `anonPieceRetrievalLastByteMs`) |
| <a id="anonCommPVerified"></a>`anonCommPVerified` | Response bytes hashed and the resulting CommP compared against the declared `pieceCid`. Inline check; no discrete event emission. Failure flips piece fetch to `failure.commp`.                                      | [`piece-retrieval.service.ts`](../../apps/backend/src/retrieval-anon/piece-retrieval.service.ts) |

> | <a id="anonCarParsed"></a>`anonCarParsed` | Fetched piece bytes are parsed as a CAR and a random sample of `ANON_RETRIEVAL_BLOCK_SAMPLE_COUNT` CIDs is selected. Only runs when the piece advertises IPFS indexing.                                              | [`car-validation.service.ts`](../../apps/backend/src/retrieval-anon/car-validation.service.ts) |

Reviewer: ```suggestion
| <a id="anonCarParsed"></a>`anonCarParsed` | Fetched piece bytes are parsed as a CAR and a random sample of `ANON_RETRIEVAL_BLOCK_SAMPLE_COUNT` CIDs is selected. Only emiitted when the piece advertises IPFS indexing.                                              | [`car-validation.service.ts`](../../apps/backend/src/retrieval-anon/car-validation.service.ts) |
```

| <a id="anonIpniVerificationStart"></a>`anonIpniVerificationStart` | Dealbot begins polling filecoinpin.contact for `<rootCid + sampled CIDs, SP>` provider records.                                                                                                                      | [`car-validation.service.ts`](../../apps/backend/src/retrieval-anon/car-validation.service.ts) |
| <a id="anonIpniVerificationComplete"></a>`anonIpniVerificationComplete` | IPNI verification finishes (all CIDs resolved to the SP, `IPNI_VERIFICATION_TIMEOUT_MS` reached, or error).                                                                                                          | [`car-validation.service.ts`](../../apps/backend/src/retrieval-anon/car-validation.service.ts) (drives `ipni_verify_ms`) |
| <a id="anonBlockFetchStart"></a>`anonBlockFetchStart` | Dealbot starts fetching the sampled CIDs via `GET {spBaseUrl}/ipfs/{cid}?format=raw`.                                                                                                                                | [`car-validation.service.ts`](../../apps/backend/src/retrieval-anon/car-validation.service.ts) |

> | <a id="anonBlockFetchComplete"></a>`anonBlockFetchComplete` | All sampled block fetches finished; each response was hash-verified against its declared CID (any non-2xx, hash mismatch, unsupported codec, or transport error counts as a failed block).                           | [`car-validation.service.ts`](../../apps/backend/src/retrieval-anon/car-validation.service.ts) |

Reviewer: Are there other termination conditions besides finishing fetching all the blocks? What about a timeout?

Author: It's really only the abort signal beyond what's listed which translates to a timeout. Note, we have a per-block-fetch timeout and per-job timeout. I've added "timeout" to the enumeration.

> | <a id="anonRetrievalCheckComplete"></a>`anonRetrievalCheckComplete` | Anonymous retrieval check terminates — successful piece fetch (plus optional CAR/IPNI/block-fetch validations) or any failure / abort. Drives the `anon_retrieval_checks` ClickHouse row and `anonRetrievalCheckMs`. | [`anon-retrieval.service.ts`](../../apps/backend/src/retrieval-anon/anon-retrieval.service.ts) (logs `anon_retrieval_completed`) |

Reviewer: I assume we also terminate after a timeout as well?

Author: yup, the `abort` is our timeout signal but I'll make it explicit 👍

| <a id="dataStorageCheckMs"></a>`dataStorageCheckMs` | Data Storage | [`uploadToSpStart`](#uploadToSpStart) | [`ipfsRetrievalIntegrityChecked`](#ipfsRetrievalIntegrityChecked) | Duration of a Data Storage check | |
| <a id="retrievalCheckMs"></a>`retrievalCheckMs` | Retrieval | Retrieval check start | [`ipfsRetrievalIntegrityChecked`](#ipfsRetrievalIntegrityChecked) | Duration of a Retrieval check | |
| <a id="dataSetCreationMs"></a>`dataSetCreationMs` | Data-Set Creation | Data-set creation uploadToSpStart | Data-set creation pieceConfirmed | Duration of one data-set creation with confirmed piece (all using `createDataSetWithPiece`) | [`deal.service.ts`](../../apps/backend/src/deal/deal.service.ts) |

> | <a id="anonPieceRetrievalFirstByteMs"></a>`anonPieceRetrievalFirstByteMs` | Anonymous Retrieval | Piece fetch start | First byte received from `/piece/{pieceCid}` | Time to first byte for anonymous piece retrievals | [`anon-retrieval.service.ts`](../../apps/backend/src/retrieval-anon/anon-retrieval.service.ts) |

Reviewer: (here and below) Per above, lets do [`eventName`](#eventName) hyperlinks into the event list we create above.

Author: Done in https://github.com/FilOzone/dealbot/pull/487/changes/2850db821e2a340a28c4c3b0a9645fbc4ffb21d2

| <a id="ipfsRetrievalHttpResponseCode"></a>`ipfsRetrievalHttpResponseCode` | Data Storage, Retrieval | [`ipfsRetrievalLastByteReceived`](#ipfsRetrievalLastByteReceived) | `200`, `500`, `2xxSuccess`, `4xxClientError`, `5xxServerError`, `otherHttpStatusCodes`, `failure` | [`retrieval.service.ts`](../../apps/backend/src/retrieval/retrieval.service.ts) |
| <a id="retrievalStatus"></a>`retrievalStatus` | Data Storage, Retrieval | [`ipfsRetrievalIntegrityChecked`](#ipfsRetrievalIntegrityChecked) | `success`, `failure.timedout`, `failure.other` from [Data Storage Sub-status meanings](./data-storage.md#sub-status-meanings). |  |
| <a id="dataSetCreationStatus"></a>`dataSetCreationStatus` | Data-Set Creation | Not tied to an [event above](#event-list) but rather to data-set creation start (`pending`) and completion (`success`/`failure.*`) | `pending`, `success`, `failure.timedout`, `failure.other` | [`deal.service.ts`](../../apps/backend/src/deal/deal.service.ts) |

> | <a id="dataSetChallengeStatus"></a>`dataSetChallengeStatus` | Data Retention | Not tied to an [event above](#event-list) but rather to the periodic chain-checking done in the [Data Retention Check](./data-retention.md) | `success`, `failure` | [`data-retention.service.ts`](../../apps/backend/src/data-retention/data-retention.service.ts) |

Reviewer: Was it intentional to change this row?

> | <a id="pdp_provider_overdue_periods"></a>`pdp_provider_overdue_periods` | Data Retention | Emitted on every poll | Gauge value (estimated overdue periods) | [`data-retention.service.ts`](../../apps/backend/src/data-retention/data-retention.service.ts) |

Reviewer: Intended to change? What do we mean by "emitted on every poll". Wasn't the previous description ore specific/accurate?

> | <a id="anonRetrievalStatus"></a>`anonRetrievalStatus` | Anonymous Retrieval | After piece fetch completes (or on abort) | `success` (HTTP 2xx **and** CommP matches), `failure.http`, `failure.commp` (HTTP 2xx but bytes hashed to a different CID), `failure.aborted`, `failure.no_piece`. | [`anon-retrieval.service.ts`](../../apps/backend/src/retrieval-anon/anon-retrieval.service.ts) |

Reviewer: ```suggestion
| <a id="anonPieceRetrievalStatus"></a>`anonPieceRetrievalStatus` | Anonymous Retrieval | After piece fetch completes (or on abort) | `success` (HTTP 2xx **and** CommP matches), `failure.http`, `failure.commp` (HTTP 2xx but bytes hashed to a different CID), `failure.timedoout`, `failure.no_piece`. | [`anon-retrieval.service.ts`](../../apps/backend/src/retrieval-anon/anon-retrieval.service.ts) |
```

Given we're calling this whole check "anon retrieval", and this is just about the piece retrieval portion, I think this requires some extra terminlogoy.
(I realize this rename involves some code and other docs changes).

Also, I see we are introducing the word "aborted" in our docs and in these status, but it sees like it's the same as "timedout". Assuming it isn't different, lets use the same term to be consistent. If it's different, lets get it explained.

Like above, it would be ideal to be hyperlinking to something into the event list.

Author: The rename wasn't too bad: https://github.com/FilOzone/dealbot/pull/487/changes/a4f0b38fdb789de01d1dbff7e8977434320c0008

I think it's reasonable to interpret an abort signal as a timeout because it'll come from the job scheduler which will in fact only send such signal in case of a timeout: https://github.com/FilOzone/dealbot/pull/487/changes/1a32373e29e7868c39f19d68ded25e3d238b9858

> | <a id="anonPieceHttpResponseCode"></a>`anonPieceHttpResponseCode` | Anonymous Retrieval | After piece fetch completes | `200`, `500`, `2xxSuccess`, `4xxClientError`, `5xxServerError`, `otherHttpStatusCodes`, `failure` (same classifier as [`ipfsRetrievalHttpResponseCode`](#ipfsRetrievalHttpResponseCode)) | [`anon-retrieval.service.ts`](../../apps/backend/src/retrieval-anon/anon-retrieval.service.ts) |

Reviewer: ```suggestion
| <a id="anonPieceHttpResponseCode"></a>`anonPieceHttpResponseCode` | Anonymous Retrieval | After piece fetch completes | Same as [`ipfsRetrievalHttpResponseCode`](#ipfsRetrievalHttpResponseCode). | [`anon-retrieval.service.ts`](../../apps/backend/src/retrieval-anon/anon-retrieval.service.ts) |
```

| <a id="anonCarParseStatus"></a>`anonCarParseStatus` | Anonymous Retrieval | After CAR validation runs (skipped when piece fetch failed or piece is not IPFS-indexed) | `parseable`, `not_parseable` | [`anon-retrieval.service.ts`](../../apps/backend/src/retrieval-anon/anon-retrieval.service.ts) |

> | <a id="anonIpniStatus"></a>`anonIpniStatus` | Anonymous Retrieval | After CAR validation runs, **or** when piece fetch failed (records `skipped`) | `valid`, `invalid`, `skipped`, `error` | [`anon-retrieval.service.ts`](../../apps/backend/src/retrieval-anon/anon-retrieval.service.ts) |

Reviewer: Do we nead to treat anonIpniStatus as special? Or should we just piggy back on `discoverabilityStatus`?

Also, do we want to use different semantics than we do for data storage / retrieval checks above? Do we lean into `success, failure.timedout, failure.other`, etc.?

Author: How would the states map?

- `valid` -> `success`
- `invalid` -> actually doesn't make sense (I'll refactor)
- `skipped` -> ? (this cannot be expressed with either failure or success but I'd like to record the fact that we didn't perform an IPNI check in Clickhouse)
- `error` -> `failure.other` or `failure.timedout`

Reviewer: Yeah, your mapping seems good to me:

valid -> success
skipped -> skipped (new)
error -> failure.other or failure.timedout seems good to me

> | <a id="anonBlockFetchStatus"></a>`anonBlockFetchStatus` | Anonymous Retrieval | After block-fetch sampling runs, **or** when piece fetch failed (records `skipped`) | `valid`, `invalid`, `skipped`, `error` | [`anon-retrieval.service.ts`](../../apps/backend/src/retrieval-anon/anon-retrieval.service.ts) |

Reviewer: Similar, rather than coming up with `valid`, `invalid`, `skipped`, `error`, do we use success, failure.timedout, failure.other, etc.

Author: Again how would they map?

- `valid` -> `success`
- `invalid` -> `failure.other`?
- `skipped` -> ?
- `error` -> `failure.other` or `failure.timedout`

I think it'd make sense to distinguish the case where we downloaded invalid blocks vs. a network error.

Author: or am I free to introduce new `failure.XYZ` values? Would I need to document them somewhere else?

Reviewer: You're free to introcue new `failure.XYZ`.

In terms of documenting statuses, lets document in anon-retrievals.md and add a note in the table like we have elsewhere with "success, failure.timedout, failure.other from [Data Storage Sub-status meanings](https://github.com/FilOzone/dealbot/blob/main/docs/checks/data-storage.md#sub-status-meanings)."

| <a id="dataSetChallengeStatus"></a>`dataSetChallengeStatus` | Data Retention | Emitted on each [Data Retention Check](./data-retention.md) poll when a provider's confirmed proving-period totals advance (strictly positive deltas). Unit: **challenges** (period delta × `CHALLENGES_PER_PROVING_PERIOD = 5`). | `success` (challenges in successfully-proven periods), `failure` (challenges in faulted periods) | [`data-retention.service.ts`](../../apps/backend/src/data-retention/data-retention.service.ts) |
| <a id="pdp_provider_estimated_overdue_periods"></a>`pdp_provider_estimated_overdue_periods` | Data Retention | Emitted on every [Data Retention Check](./data-retention.md) poll for every successfully processed provider. | Gauge value in proving periods (non-negative integer) | [`data-retention.service.ts`](../../apps/backend/src/data-retention/data-retention.service.ts) |

> | <a id="anonPieceRetrievalStatus"></a>`anonPieceRetrievalStatus` | Anonymous Retrieval | After piece fetch completes (or on abort) | `success` (HTTP 2xx **and** CommP matches), `failure.http`, `failure.commp` (HTTP 2xx but bytes hashed to a different CID), `failure.timedout`, `failure.no_piece`. | [`anon-retrieval.service.ts`](../../apps/backend/src/retrieval-anon/anon-retrieval.service.ts) |

Reviewer: Here and below, it would be good to flush out the status and their meanings like we have with https://github.com/probe-lab/dealbot/blob/48ecadb21bb9eecf73ea2e7c8209a007ff4d4050/docs/checks/data-storage.md#sub-status-meanings . I don't want their to be ambiguity what these mean.

Author: Done in https://github.com/FilOzone/dealbot/pull/487/changes/2850db821e2a340a28c4c3b0a9645fbc4ffb21d2
